src/journal/journal-file.c

```diff
@@ -2062,9 +2067,14 @@ static int generic_array_bisect(
 
                                 p = le64toh(array->entry_array.items[i]);
                                 if (p <= 0)
-                                        return -EBADMSG;
-
-                                r = test_object(f, p, needle);
+                                        r = -EBADMSG;
+                                else
+                                        r = test_object(f, p, needle);
+                                if (r == -EBADMSG) {
+                                        log_debug_errno(r, "Encountered invalid entry while bisecting, cutting algorithm short. (2)");
+                                        right = n = i;
+                                        continue;
+                                }
```

Comment:
test_object() might return -EBADMSG too. Or in other words: p <= tells us the offset is bogus, and test_object() tells us whether the object behind the offset is bogus. For us the distinction doesn't matter much, in either case, we want to cut the bisection short.